### PR TITLE
Add OCP web console enrichments (ocp4.3 or after)

### PR DESCRIPTION
**Pull request by myeung18**
Signed-off-by: Marco Yeung <myeung@redhat.com>

### CODE REVIEW

File: pkg/controller/initializer.go [hunk 1]
```diff
@@ -13,6 +14,7 @@ func init() {
 	addManager := func(mgr manager.Manager) error {
 		k8sService := GetInstance(mgr)
 		reconciler := kieapp.Reconciler{Service: &k8sService}
+		webconsole.LoadWebConsoleEnrichmentUI()
```
Comment (bmozaffa) [inline pkg/controller/initializer.go]: It seems to me that this is similar to creating our operator UI. We do that [here](https://github.com/kiegroup/kie-cloud-operator/blob/master/pkg/controller/kieapp/kieapp_controller.go#L58-L60), so I would say this line would go right next to it. @tchughesiv , wdyt?

File: pkg/controller/kieapp/webconsole/webconsole.go [hunk 2]
```diff
@@ -0,0 +1,102 @@
+package webconsole
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole/webconsole.go]: Maybe just call it olm?

File: pkg/controller/kieapp/webconsole/webconsole.go [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	files, err := loadFilesOnlyWithBox("consoleyamlsamples", "noop", "noop")
+	if err != nil {
+		log.Error(err, "yaml stings not loaded")
+	} else {
+		_, err = applyMultipleWebConsoleYamls(files)
+		if err != nil {
+			log.Error(err, "web console yamlsamples not successfully applied")
+		}
+	}
+	files, err = loadFilesOnlyWithBox("webconsole", "noop", "noop")
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole/webconsole.go]: This operator only needs sample yamls in this code. The only other thing is the link, but it won't be based on a yaml and will have to be separately figured out based on the deployed object, so please take out the other stuff from here and just focus on creating the yaml samples.

Comment (bmozaffa) [reply pkg/controller/kieapp/webconsole/webconsole.go]: Given the above, chance the function name to something like "CreateSampleYAMLs"

File: pkg/controller/kieapp/webconsole/webconsole.go [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+		log.Error(err, "yaml stings not loaded")
+	} else {
+		_, err = applyMultipleWebConsoleYamls(files)
+		if err != nil {
+			log.Error(err, "web console enrichments not successfully applied")
+		}
+	}
+}
+
+func applyMultipleWebConsoleYamls(yamlString []string) (map[int]string, error) {
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole/webconsole.go]: I'm not sure why we need all of this. It's 1 line to create an object, and an extra 3-4 lines to handle the error. For example, see [here](https://github.com/kiegroup/kie-cloud-operator/blob/master/pkg/controller/kieapp/kieapp_controller.go#L682-L686).

You just need to create an object type based on the existing yaml structure that the code can then create.

**Comment by ruromero** [comment]
Consider depending on this PR https://github.com/kiegroup/kie-cloud-operator/pull/386

File: pkg/controller/kieapp/webconsole/webconsole.go (hunk 2 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole/webconsole.go]: probably not as all these customizations are meant for OCP Container platform and independent of the olm.

File: pkg/controller/kieapp/webconsole/webconsole.go (hunk 3 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole/webconsole.go]: modified.

File: pkg/controller/initializer.go (hunk 1 above)
Comment (myeung18) [reply pkg/controller/initializer.go]: moved. 
it looks like a proper place to put the yaml sample creation method call.

File: pkg/controller/kieapp/webconsole/webconsole.go (hunk 4 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole/webconsole.go]: These are all here is meant to deal with a bunch of yaml files passed in, and we don't know exactly what type they are (unless we read the file and get the kind), they could be yaml sample, consolelink, consoleexternaloglink etc.
All these are used to find the object type and then delegate to the right creator to create.

Comment (bmozaffa) [reply pkg/controller/kieapp/webconsole/webconsole.go]: Yes, I am not aware of any different type of yaml that needs to be handled in this code, or a situation where the caller is passing in a yaml and doesn't know what it is. Let's make this quick and simple, to handle the only types we need to create here, which is CR samples.

File: pkg/controller/kieapp/webconsole/webconsole.go (hunk 2 above)
Comment (bmozaffa) [reply pkg/controller/kieapp/webconsole/webconsole.go]: Sure, I guess webconsole is fine

**Comment by tchughesiv** [comment]
please rebase

File: pkg/controller/kieapp/webconsole/webconsole.go (hunk 4 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole/webconsole.go]: That means we don't need the factor and creator. Modified.

File: pkg/controller/kieapp/webconsole/webconsole.go [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+type CustomResourceDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+}
+
+var log = logs.GetLogger("kieapp.webconsole")
+
+func CreateSampleYAMLs() {
+	log.Info("Loading Console yaml samples.")
+	files, err := loadFilesOnlyWithBox("consoleyamlsamples")
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole/webconsole.go]: This should use actual files in `deploy/crs`. I'm not sure what this currently does or denotes.

File: pkg/controller/kieapp/webconsole/webconsole.go [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+		if err != nil {
+			resMap[idx] = err.Error()
+		} else {
+			resMap[idx] = "Applied"
+		}
+	}
+	return resMap, nil
+}
+
+func create(yamlStr string) (bool, error) {
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole/webconsole.go]: There is a whole lot of complexity here, and I'm not sure why. The operator already has a `writer` that is used throughout the project to create resources.

File: pkg/controller/kieapp/webconsole/webconsole.go (hunk 5 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole/webconsole.go]: currently for testing, now moved to crs folder.

File: pkg/controller/kieapp/webconsole/webconsole.go (hunk 6 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole/webconsole.go]: Should've used the reconciler from controller. Modified.
now only webconsole.go is left.

File: deploy/crs/kieapp_rhdm_trial.yaml [hunk 7]
```diff
@@ -2,5 +2,10 @@ apiVersion: app.kiegroup.org/v2
 kind: KieApp
 metadata:
   name: rhdm-trial
+  annotations:
+    consolekind: ConsoleYAMLSample
+    consolename: kieapp-rhdm-trial
+    consoledesc: KieApp rhdm trial
+    consoletitle: KieApp rhdm trial
```
Comment (bmozaffa) [inline deploy/crs/kieapp_rhdm_trial.yaml]: Please use camel case, e.g., consoleTitle

File: deploy/crs/kieapp_rhdm_trial.yaml [hunk 8]
```diff
@@ -2,5 +2,10 @@ apiVersion: app.kiegroup.org/v2
 kind: KieApp
 metadata:
   name: rhdm-trial
+  annotations:
+    consolekind: ConsoleYAMLSample
```
Comment (bmozaffa) [inline deploy/crs/kieapp_rhdm_trial.yaml]: Why is this necessary?

File: pkg/controller/kieapp/webconsole.go [hunk 9]
```diff
@@ -0,0 +1,141 @@
...
+	TargetNamespace   = "openshift-console"
+	ConsoleYAMLSample = "ConsoleYAMLSample"
+)
+
+type CustomResourceDefinition struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+}
+
+func CreateConsoleYAMLSamples(r *Reconciler) {
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole.go]: This should be package private, so start with lowercase: `create`

File: pkg/controller/kieapp/webconsole.go [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	TargetNamespace   = "openshift-console"
+	ConsoleYAMLSample = "ConsoleYAMLSample"
+)
+
+type CustomResourceDefinition struct {
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole.go]: Shouldn't need this

File: pkg/controller/kieapp/webconsole.go [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+		} else {
+			resMap[idx] = "Applied"
+		}
+	}
+	return resMap, nil
+}
+
+func create(yamlStr string, r *Reconciler) (bool, error) {
+	obj := &CustomResourceDefinition{}
+	err := yaml.Unmarshal([]byte(yamlStr), obj)
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole.go]: Our unit tests or somewhere should have the right way of reading our CR files

File: pkg/controller/kieapp/webconsole.go [hunk 12]
```diff
@@ -0,0 +1,141 @@
...
+				APIVersion: obj.TypeMeta.APIVersion,
+				Kind:       obj.TypeMeta.Kind,
+			},
+			Title:       consolev1.ConsoleYAMLSampleTitle(title),
+			Description: consolev1.ConsoleYAMLSampleDescription(desc),
+			YAML:        consolev1.ConsoleYAMLSampleYAML(yamlStr),
+			Snippet:     snippet,
+		},
+	}
+	err = r.Service.Create(context.TODO(), yamlSample)
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole.go]: Now thinking about what can be generic, this line won't be portable. I think we should make a simple util function that takes the CR and returns a `ConsoleYAMLSample`, and let the operator create it.

File: deploy/crs/kieapp_rhdm_trial.yaml (hunk 8 above)
Comment (myeung18) [reply deploy/crs/kieapp_rhdm_trial.yaml]: It is not needed. However, doing gives us flexibility to selective apply certain CRs (ie:apply only those have this 'consoleKind'), instead of apply the CRs in the entire folder.

Comment (bmozaffa) [reply deploy/crs/kieapp_rhdm_trial.yaml]: The caller will be passing a yaml to create a console yaml sample. They will want it created and we shouldn't be second-guessing their call. In fact, we should use defaults if no annotations exist at all.

File: deploy/crs/kieapp_rhdm_trial.yaml (hunk 7 above)
Comment (myeung18) [reply deploy/crs/kieapp_rhdm_trial.yaml]: modified

File: pkg/controller/kieapp/webconsole.go (hunk 9 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole.go]: modified

File: pkg/controller/kieapp/webconsole.go (hunk 10 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole.go]: removed

File: pkg/controller/kieapp/webconsole.go (hunk 11 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole.go]: modified

File: pkg/controller/kieapp/webconsole.go (hunk 12 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole.go]: modified

File: pkg/controller/kieapp/shared/utils.go [hunk 13]
```diff
@@ -174,3 +180,52 @@ func GetNamespacedName(object metav1.Object) types.NamespacedName {
...
 }
+
+const (
+	TargetNamespace   = "openshift-console"
+	ConsoleYAMLSample = "ConsoleYAMLSample"
+)
+
+func GetConsoleYAMLSample(res resource.KubernetesResource) (*consolev1.ConsoleYAMLSample, error) {
+	kind := res.GetAnnotations()["consoleKind"]
+	if kind != ConsoleYAMLSample {
```
Comment (bmozaffa) [inline pkg/controller/kieapp/shared/utils.go]: As mentioned, we should remove this. All the annotations should be optional.

File: pkg/controller/kieapp/shared/utils.go [hunk 14]
```diff
@@ -174,3 +180,52 @@ func GetNamespacedName(object metav1.Object) types.NamespacedName {
...
+		if err != nil {
+			return nil, errors.New("Unable to parse snippet to boolean.")
+		}
+		snippet = tmp
+	}
+	title, _ := res.GetAnnotations()["consoleTitle"]
+	desc, _ := res.GetAnnotations()["consoleDesc"]
+	name, _ := res.GetAnnotations()["consoleName"]
+
+	res.SetAnnotations(make(map[string]string))
```
Comment (bmozaffa) [inline pkg/controller/kieapp/shared/utils.go]: They may have their own annotations that need to be in the CR. Only remove our 3 known annotations.

File: pkg/controller/kieapp/shared/utils.go [hunk 15]
```diff
@@ -174,3 +180,51 @@ func GetNamespacedName(object metav1.Object) types.NamespacedName {
 		Namespace: object.GetNamespace(),
 	}
 }
+
+func GetConsoleYAMLSample(res resource.KubernetesResource) (*consolev1.ConsoleYAMLSample, error) {
+	snippetStr := res.GetAnnotations()["consoleSnippet"]
```
Comment (bmozaffa) [inline pkg/controller/kieapp/shared/utils.go]: I'm not clear on what this does and what purpose it serves. So it's ok to leave it out, but if the operator passes in a yaml with consoleSnippet set to false, we return an error instead of a ConsoleYAMLSample? Not sure that's necessary, or why an operator would pass it in, if they've marked as false / not a snippet. I would just remove this whole part.

File: pkg/controller/kieapp/shared/utils.go [hunk 16]
```diff
@@ -174,3 +180,51 @@ func GetNamespacedName(object metav1.Object) types.NamespacedName {
...
+	if snippetStr != "" {
+		tmp, err := strconv.ParseBool(snippetStr)
+		if err != nil {
+			return nil, errors.New("Unable to parse snippet to boolean.")
+		}
+		snippet = tmp
+	}
+
+	title, _ := res.GetAnnotations()["consoleTitle"]
+	desc, _ := res.GetAnnotations()["consoleDesc"]
```
Comment (bmozaffa) [inline pkg/controller/kieapp/shared/utils.go]: I haven't seen how things should look in the web console... so I take it that title and description is optional? Fine if it is, just checking.

File: pkg/controller/kieapp/shared/utils.go [hunk 17]
```diff
@@ -174,3 +180,51 @@ func GetNamespacedName(object metav1.Object) types.NamespacedName {
...
+	name, _ := res.GetAnnotations()["consoleName"]
+	if name == "" {
+		name = res.GetName() + "-yamlsample"
+	}
+
+	curannos := res.GetAnnotations()
```
Comment (bmozaffa) [inline pkg/controller/kieapp/shared/utils.go]: How about giving this a better name (maybe just `annotations`?), moving this further up, and using it to retrieve the values too?

File: pkg/controller/kieapp/shared/utils.go [hunk 18]
```diff
@@ -174,3 +180,51 @@ func GetNamespacedName(object metav1.Object) types.NamespacedName {
...
+				Kind:       res.GetObjectKind().GroupVersionKind().Kind,
+			},
+			Title:       consolev1.ConsoleYAMLSampleTitle(title),
+			Description: consolev1.ConsoleYAMLSampleDescription(desc),
+			YAML:        consolev1.ConsoleYAMLSampleYAML(string(data)),
+			Snippet:     snippet,
+		},
+	}
+	return yamlSample, nil
+}
```
Comment (bmozaffa) [inline pkg/controller/kieapp/shared/utils.go]: I think this looks good otherwise. Please find a good folder and package name for it in operator-utils and open a PR there. Also create a unit test for it there, and let's prepare to merge it.

File: pkg/controller/kieapp/shared/utils.go (hunk 15 above)
Comment (myeung18) [reply pkg/controller/kieapp/shared/utils.go]: Since now we by default try our best to create the ConsoleYAMLSample, this error return is not necessary, and we should just give a default value (false: not a snippet).   But we should still have consoleSnippet to allow user to turn the entire yaml as snippet: set to true.
I'd do like this:
	var snippet bool = false
          .....
		tmp, err := strconv.ParseBool(snippetStr)
		if err == nil {
			snippet = tmp
		}

File: pkg/controller/kieapp/shared/utils.go (hunk 16 above)
Comment (myeung18) [reply pkg/controller/kieapp/shared/utils.go]: They are optional.

File: pkg/controller/kieapp/shared/utils.go (hunk 17 above)
Comment (myeung18) [reply pkg/controller/kieapp/shared/utils.go]: modified

File: pkg/controller/kieapp/shared/utils.go (hunk 18 above)
Comment (myeung18) [reply pkg/controller/kieapp/shared/utils.go]: will do.  first, will make this kie-cloud pr working first.

File: pkg/controller/kieapp/shared/utils.go (hunk 15 above)
Comment (bmozaffa) [reply pkg/controller/kieapp/shared/utils.go]: I don't understand!

Comment (myeung18) [reply pkg/controller/kieapp/shared/utils.go]: by default this yaml file is not a snippet.  If user wants to mark it as snippet, he gives the consoleSnippet: true. But if the string value has typo: not parsable, we keep it as false.

Comment (bmozaffa) [reply pkg/controller/kieapp/shared/utils.go]: oh, I misunderstood this whole thing. What is a snippet? That means the yaml is not your whole CR? Is there a difference in behavior, perhaps in that it doesn't completely replace your yaml when clicked?

Comment (myeung18) [reply pkg/controller/kieapp/shared/utils.go]: Correct, different behavior, doesn't completely replace when clicked.
another tab will be created and named "Snippet", normal Yamls have "Samples" tab.

File: pkg/controller/kieapp/shared/utils.go [hunk 19]
```diff
@@ -174,3 +180,50 @@ func GetNamespacedName(object metav1.Object) types.NamespacedName {
...
 	}
 }
+
+func GetConsoleYAMLSample(res resource.KubernetesResource) (*consolev1.ConsoleYAMLSample, error) {
+	annotations := res.GetAnnotations()
+	snippetStr := annotations["consoleSnippet"]
+	var snippet bool = false
+	if snippetStr != "" {
+		tmp, err := strconv.ParseBool(snippetStr)
+		if err == nil {
```
Comment (bmozaffa) [inline pkg/controller/kieapp/shared/utils.go]: Nothing wrong with this, but you can also do:

```
if tmp, err := strconv.ParseBool(snippetStr); err == nil {
   snippet = tmp
}
```

File: pkg/controller/kieapp/webconsole/webconsole.go (hunk 5 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole/webconsole.go]: done

File: deploy/crs/kieapp_rhdm_authoring.yaml [hunk 20]
```diff
@@ -2,5 +2,6 @@ apiVersion: app.kiegroup.org/v2
 kind: KieApp
 metadata:
   name: rhdm-authoring
+  annotations:
```
Comment (bmozaffa) [inline deploy/crs/kieapp_rhdm_authoring.yaml]: Why this? Did you find annotations needed to be there or something?

File: pkg/controller/kieapp/webconsole.go [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+		return false, err
+	}
+	err = r.Service.Create(context.TODO(), yamlSample)
+	if err != nil {
+		return false, err
+	}
+	return true, nil
+}
+
+func loadFiles(boxname string, path ...string) ([]string, []string, error) {
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole.go]: Why is a boxname needed? Why not just hardcode to something on line 75?

File: deploy/crs/kieapp_rhdm_authoring.yaml (hunk 20 above)
Comment (myeung18) [reply deploy/crs/kieapp_rhdm_authoring.yaml]: cleaned, I need more vigilance.

File: pkg/controller/kieapp/webconsole.go (hunk 21 above)
Comment (myeung18) [reply pkg/controller/kieapp/webconsole.go]: that way looks better, modified.

File: go.mod [hunk 22]
```diff
@@ -15,7 +15,7 @@ require (
 	github.com/heroku/docker-registry-client v0.0.0-20190909225348-afc9e1acc3d5
 	github.com/imdario/mergo v0.3.8
 	github.com/openshift/api v3.9.1-0.20190924102528-32369d4db2ad+incompatible
-	github.com/openshift/client-go v0.0.0-20200116152001-92a2713fa240
+	github.com/openshift/client-go v3.9.0+incompatible
```
Comment (bmozaffa) [inline go.mod]: This version name is fishy! Is this right? No adverse effects? @tchughesiv ?

File: pkg/controller/kieapp/webconsole.go [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+	"context"
+	"errors"
+	"fmt"
+	"github.com/RHsyseng/operator-utils/pkg/utils/openshift"
+	"github.com/ghodss/yaml"
+	"github.com/gobuffalo/packr/v2"
+	api "github.com/kiegroup/kie-cloud-operator/pkg/apis/app/v2"
+)
+
+func CreateConsoleYAMLSamples(r *Reconciler) {
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole.go]: I think this file will look much better if you combine these 4 into one function.

Comment (myeung18) [reply pkg/controller/kieapp/webconsole.go]: modified

File: pkg/controller/kieapp/webconsole.go [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+		resMap[idx] = "Applied"
+	}
+
+	for k, v := range resMap {
+		log.Infof("yaml file: %s %s ", names[k], v)
+	}
+}
+
+//used by packr
+func convertYamlToBinary() {
```
Comment (bmozaffa) [inline pkg/controller/kieapp/webconsole.go]: Is this old / should be removed?

Comment (myeung18) [reply pkg/controller/kieapp/webconsole.go]: redundant.

File: go.mod (hunk 22 above)
Comment (tchughesiv) [reply go.mod]: there should be no issues with this as the important line in go.mod for this library is this one in the replace section - https://github.com/kiegroup/kie-cloud-operator/blob/ad198e7e68bafeb5e477dbfe1e45df40cbb35520/go.mod#L48

which forces this in go.sum - https://github.com/kiegroup/kie-cloud-operator/blob/ad198e7e68bafeb5e477dbfe1e45df40cbb35520/go.sum#L668

that said, i'm not sure why this line is changing with this PR... maybe carryover from he operator-utils library change?

Comment (bmozaffa) [reply go.mod]: Any thoughts, Marco?

Comment (myeung18) [reply go.mod]: I also think the version at the replace section will take the precedence over that version at the require section. 
That line seem to be generated by go mod as we didn't specify the exact version when operator-utils was converted to use go mod from dep. Now it is carried forward.
So, I think we should be ok.

Btw, why do we use a lot of replace sections given that we know exactly what version we want?
If an exact version was listed in operator-utils' require section, could we have less this  type of version concern?

Comment (bmozaffa) [reply go.mod]: Good question, Marco. I'm not sure and maybe Tommy is better placed to answer this question. I suppose we want to be thorough and avoid this project breaking in case operator-utils uses a different version. In other words, avoid tight coupling to that project.
